src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs

```diff
@@ -430,15 +432,83 @@ private void RunTestCleanupMethod(object classInstance, TestResult result)
+                        cleanupStackTrace.Append(Environment.NewLine);
+                    }
+                }
+
+                Exception realException = ex.GetInnerExceptionOrDefault();
+
+                // special case UnitTestAssertException to trim off part of the stack trace
+                if (realException is UnitTestAssertException)
+                {
+                    cleanupOutcome =
+                        realException is AssertInconclusiveException ? UTF.UnitTestOutcome.Inconclusive : UTF.UnitTestOutcome.Failed;
+                    cleanupError.Append(StackTraceHelper.GetExceptionMessage(realException));
+
+                    StackTraceInformation realExceptionStackTraceInfo = StackTraceHelper.GetStackTraceInformation(realException);
+                    if (realExceptionStackTraceInfo != null)
+                    {
+                        cleanupStackTrace.Append(realExceptionStackTraceInfo.ErrorStackTrace);
+                        cleanupStackTraceInfo = cleanupStackTraceInfo ?? realExceptionStackTraceInfo;
+                    }
+                }
+                else
+                {
+                    cleanupOutcome = UTF.UnitTestOutcome.Failed;
+                    string message = testCleanupMethod != null ?
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Resource.UTA_CleanupMethodThrows,
+                            this.TestClassName,
+                            testCleanupMethod?.Name,
+                            realException.GetType().ToString(),
+                            StackTraceHelper.GetExceptionMessage(realException)) :
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Resource.UTA_CleanupMethodThrowsGeneralError,
+                            this.TestClassName,
+                            StackTraceHelper.GetExceptionMessage(realException));
+
+                    cleanupError.Append(message);
+
+                    StackTraceInformation realExceptionStackTraceInfo = StackTraceHelper.GetStackTraceInformation(realException);
+                    if (realExceptionStackTraceInfo != null)
+                    {
+                        cleanupStackTrace.Append(realExceptionStackTraceInfo.ErrorStackTrace);
+                        cleanupStackTraceInfo = cleanupStackTraceInfo ?? realExceptionStackTraceInfo;
+                    }
+                }
+
+                UTF.UnitTestOutcome outcome = testFailureException == null ? cleanupOutcome : UnitTestOutcomeHelper.GetMoreImportantOutcome(cleanupOutcome, result.Outcome);
+                StackTraceInformation finalStackTraceInfo = cleanupStackTraceInfo != null ?
+                                new StackTraceInformation(
+                                    cleanupStackTrace.ToString(),
+                                    cleanupStackTraceInfo.ErrorFilePath,
```

Comment:
Should this be pointing to the Test Method failure instead? Cause that could be a reason for the cleanup failure?